python/ray/train/worker_group.py

```diff
@@ -279,6 +286,9 @@ def execute_single(self, worker_index: int, func: Callable[..., T], *args,
     def remove_workers(self, worker_indexes: List[int]):
```

Comment:
Now with placement groups the semantics here get a little tricky since we can’t add workers without first removing them if placement groups is enabled.

Do you think we should do either of the following?

1. Make remove_workers and add_workers private, and only expose a single replace_workers API. This will ensure that remove and add are called atomically (except for the initial creation) and your placement group will always be “full”.
2. Keep track of the free bundles in the placement group and if add_workers is called without free resources in the placement group, then raise an error.

Reply:
Great point, my initial thoughts are:
1. This is reasonable for the current system, but I don't think it should be a requirement of the `WorkerGroup`/`ActorGroup`. With the presented API we don't really make any assumptions on the size of the placement group.
2. This could get tricky keeping track of what a "free" bundle is (if worker:bundle is 1:1 this is easier but not sure if this makes sense). Would prefer to leave this up to placement groups to be responsible for this and raise the error when this happens (https://github.com/ray-project/ray/issues/18018). 

I do actually think https://github.com/ray-project/ray/issues/18524 is the preferred solution for Ray Train and placement groups are just a workaround to achieve this functionality, so it doesn't seem right to me to build `WorkerGroup` around placement group semantics.

Reply:
I added a [comment](https://github.com/ray-project/ray/pull/20091/commits/79478c3237d25983076ca0114653b617845eb562) to `add_workers` to warn any users about this. 

- Even prior to this change, the same undefined behavior would be present if `add_workers` is called within a parent placement group or if the cluster is full.
- The only usecase in `TensorflowBackend` actually has some custom logic in between `remove_workers` and `add_workers`.